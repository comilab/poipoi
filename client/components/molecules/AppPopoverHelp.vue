<template lang="pug">
VPopover.inline-block(
  placement="top",
  trigger="click hover",
  v-bind="$attrs",
  :popper-options="popperOptions"
)
  AppButtonHelp.text-blue-600.cursor-pointer.focus_outline-none.ml-2
  template(#popover)
    .text-sm
      slot
</template>

<script lang="ts">
import { defineComponent } from 'nuxt-composition-api'
import { VPopover } from 'v-tooltip'

import AppButtonHelp from '~/components/atoms/AppButtonHelp.vue'

export default defineComponent({
  components: {
    VPopover,
    AppButtonHelp
  },
  setup () {
    const popperOptions = {
      modifiers: {
        // 横幅が小さい時に配置がずれるので無理やり修正
        modifyLeftPosition: {
          order: 890,
          enabled: true,
          fn: (data: any) => {
            const matches = data.styles.transform.match(/translate3d\((-?[0-9]+?)(px.*?)\)/)
            if (matches) {
              const [left, others] = matches.slice(1)
              const newLeft = Math.max(0, parseInt(left))
              data.styles.transform = `translate3d(${newLeft}${others})`
            }
            return data
          }
        }
      }
    }

    return {
      popperOptions
    }
  }
})
</script>
