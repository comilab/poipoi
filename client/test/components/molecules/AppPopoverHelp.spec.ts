import { render } from '@testing-library/vue'
import { defineComponent } from 'nuxt-composition-api'

import AppPopoverHelp from '~/components/molecules/AppPopoverHelp.vue'

describe('components/molecules/AppPopoverHelp', () => {
  let options: any
  let popoverProps: any

  beforeEach(() => {
    popoverProps = {}
    options = {
      stubs: {
        FontAwesomeIcon: true,
        VPopover: defineComponent({
          props: ['placement', 'trigger', 'popperOptions'],
          setup (props) {
            popoverProps = props
          },
          template: '<div><slot name="popover" /></div>'
        })
      },
      slots: {}
    }
  })

  test('renders correctly', () => {
    const { container } = render(AppPopoverHelp, options)

    expect(container).toMatchSnapshot()
  })

  describe('slots', () => {
    describe('default', () => {
      test('popover内に表示', () => {
        options.slots.default = 'slot-default'
        const { getByText } = render(AppPopoverHelp, options)

        getByText('slot-default')
      })
    })
  })

  describe('VPopover', () => {
    test('placement = top', () => {
      render(AppPopoverHelp, options)

      expect(popoverProps.placement).toBe('top')
    })

    test('trigger = "click hover"', () => {
      render(AppPopoverHelp, options)

      expect(popoverProps.trigger).toBe('click hover')
    })

    describe('popperOptions', () => {
      test('order = 890', () => {
        render(AppPopoverHelp, options)

        expect(popoverProps.popperOptions.modifiers.modifyLeftPosition.order).toBe(890)
      })

      test('enable = true', () => {
        render(AppPopoverHelp, options)

        expect(popoverProps.popperOptions.modifiers.modifyLeftPosition.enabled).toBeTruthy()
      })

      test('translate3dのleftが0未満にならない', () => {
        render(AppPopoverHelp, options)
        const result = popoverProps.popperOptions.modifiers.modifyLeftPosition.fn({
          styles: {
            transform: 'translate3d(-1px,0)'
          }
        })

        expect(result.styles.transform).toBe('translate3d(0px,0)')
      })
    })
  })
})
