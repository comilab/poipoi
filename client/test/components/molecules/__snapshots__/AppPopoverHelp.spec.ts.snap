// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppPopoverHelp renders correctly 1`] = `
<div>
  <div
    class="inline-block"
  >
    <div
      class="text-sm"
    />
  </div>
</div>
`;
